internal/distributed/connection_manager.go

```diff
@@ -0,0 +1,309 @@
+	case typeutil.QueryNodeRole:
+		//cm.queryNodes[session.ServerID] = grpcquerynodeclient.NewClient(connection)
+	}
+}
+
+func (cm *ConnectionManager) buildConnections(session *sessionutil.Session) {
+	task := newBuildClientTask(session, cm.notify)
+	cm.buildTasks[session.ServerID] = task
+	task.Run()
+}
+
+func (cm *ConnectionManager) removeTask(id int64) {
+	cm.taskMu.Lock()
+	task, ok := cm.buildTasks[id]
+	if ok {
+		task.Stop()
+		delete(cm.buildTasks, id)
+	}
+	cm.taskMu.Unlock()
+}
+
+type buildConnectionstate int
+
+const (
+	buildConnectionstart buildConnectionstate = iota
+	buildClientRunning
+	buildConnectionsuccess
+	buildClientFailed
+)
+
+type buildClientTask struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	sess         *sessionutil.Session
+	state        buildConnectionstate
+	retryOptions []retry.Option
+
+	result *grpc.ClientConn
+	notify chan int64
+}
+
+func newBuildClientTask(session *sessionutil.Session, notify chan int64, retryOptions ...retry.Option) *buildClientTask {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &buildClientTask{
+		ctx:    ctx,
+		cancel: cancel,
+
+		sess:         session,
+		retryOptions: retryOptions,
+
+		notify: notify,
+	}
+
+}
+
+func (bct *buildClientTask) Run() {
+	bct.state = buildClientRunning
+	go func() {
+		defer bct.finish()
+		connectGrpcFunc := func() error {
+			opts := trace.GetInterceptorOpts()
+			log.Debug("Grpc connect ", zap.String("Address", bct.sess.Address))
+			conn, err := grpc.DialContext(bct.ctx, bct.sess.Address,
+				grpc.WithInsecure(), grpc.WithBlock(), grpc.WithTimeout(30*time.Second),
+				grpc.WithDisableRetry(),
+				grpc.WithUnaryInterceptor(
+					grpc_middleware.ChainUnaryClient(
+						grpc_retry.UnaryClientInterceptor(
+							grpc_retry.WithMax(3),
+							grpc_retry.WithPerRetryTimeout(time.Second*20),
+							grpc_retry.WithCodes(codes.Aborted, codes.Unavailable),
+						),
+						grpc_opentracing.UnaryClientInterceptor(opts...),
+					)),
+				grpc.WithStreamInterceptor(
+					grpc_middleware.ChainStreamClient(
+						grpc_retry.StreamClientInterceptor(
+							grpc_retry.WithMax(3),
+							grpc_retry.WithPerRetryTimeout(time.Second*20),
+							grpc_retry.WithCodes(codes.Aborted, codes.Unavailable),
+						),
+						grpc_opentracing.StreamClientInterceptor(opts...),
+					)),
+			)
+			if err != nil {
+				return err
+			}
+			bct.result = conn
+			return nil
+		}
+
+		err := retry.Do(bct.ctx, connectGrpcFunc, bct.retryOptions...)
+		if err != nil {
+			log.Debug("BuildClientTask try connect failed",
+				zap.Any("ServerName", bct.sess.ServerName), zap.Error(err))
+			bct.state = buildClientFailed
+			return
+		}
+	}()
+}
+func (bct *buildClientTask) Stop() {
+	bct.cancel()
+}
+
+func (bct *buildClientTask) finish() {
+	bct.notify <- bct.sess.ServerID
+}
+
+func (cm *ConnectionManager) checkServerName(serverName string) bool {
+	switch serverName {
```

Comment:
we can have a map[string] for all supported roles, then we only need to check whether the role name is in map.